astropy/io/ascii/tests/test_write_read.py

```diff
@@ -0,0 +1,153 @@
+
+import copy
+
+from ....extern import six
+from ....extern.six.moves import cStringIO as StringIO
+from ....tests.helper import pytest
+from ... import ascii
+from ....table import Table
+
+# importing .common sets the current directory
+#   so that test files are read from the right place
+from .common import setup_function, teardown_function
+
+# We will test that test tables that can be written (test_write.test_defs)
+#  can also be read in.
+from .test_write import test_defs
+
+debug = False
+
+# BeautifulSoup is required to *read* HTML Files
+# Check to see if the BeautifulSoup dependency is present.
+try:
+    from bs4 import BeautifulSoup, FeatureNotFound
+    HAS_BEAUTIFUL_SOUP = True
+except ImportError:
+    HAS_BEAUTIFUL_SOUP = False
+
+
+def create_reader_kwargs_from_writer_kwargs(test_write_def):
+    """Create a test definition with appropriate kwargs for a Reader."""
+    remove_col_for_read = ('formats', 'include_names', 'exclude_names',
+                           'strip_whitespace')
+
+    test_read_def = copy.deepcopy(test_write_def)
+    if 'Writer' in test_read_def['kwargs']:
+        test_read_def['kwargs']['Reader'] = test_write_def['kwargs']['Writer']
+        del test_read_def['kwargs']['Writer']
+    # Delimiter is a special case.
+    # In general, we want to keep the 'delimiter' option unchanged.
+    # But in the special case of 'delimiter=None',
+    #   we want to remove 'delimiter' from 'kwargs'
+    #   because it can confuse the Reader.
+    if 'delimiter' in test_read_def['kwargs']:
+        if test_read_def['kwargs']['delimiter'] is None:
+            del test_read_def['kwargs']['delimiter']
+    for name in remove_col_for_read:
+        if name in test_read_def['kwargs']:
+            del test_read_def['kwargs'][name]
+
+    return test_read_def
+
+
+def check_read_write_table_via_ascii(test_def, table, fast_writer):
+    """Check reading and writing using the astropy.io.ascii interface."""
+    in_out = StringIO()
+
+    test_read_def = create_reader_kwargs_from_writer_kwargs(test_def)
+    try:
+        if debug:
+            print(test_def['kwargs'])
+            print(test_read_def['kwargs'])
+        ascii.write(table, in_out, fast_writer=fast_writer,
+                    **test_def['kwargs'])
+        foo = ascii.read(in_out, **test_read_def['kwargs'])
+    except ValueError as e:  # if format doesn't have a fast writer, ignore
+        if 'not in the list of formats with fast writers' not in str(e):
+            raise e
+        return
+
+    if debug:
+        print('Expected:\n%s' % test_def['out'])
+        print('Actual:\n%s' % in_out.getvalue())
+        print('Read:\n%s' % str(foo))
+
+    assert [x.strip() for x in in_out.getvalue().strip().splitlines()] == \
+           [x.strip() for x in test_def['out'].strip().splitlines()]
+
+
+def check_read_write_table_via_table(test_def, table, fast_writer):
+    """Check reading and writing using the astropy.table.Table interface."""
```

Comment:
I was just copying the testing structure as it is in `test_write.py`.

A separate ticket to streamline table testing might be appropriate.  I'm not a particularly big fan of this "almost-the-same-thing" functions here and would certainly be happy to make this cleaner, but I don't fully understand what the expected inconsistencies might be.